src/encoding.rs

```diff
@@ -1,78 +1,384 @@
+        // shared mutable reference. A closure finalize returning the word
+        // accumulator cannot therefore be written simply.
+        let mut ws = Vec::<[u8; WORD_LENGTH]>::new();
+
+        // Writes the given bytes to the current word `w` starting at the current
+        // position `pos`, overflowing into new words if the number of bytes to be
+        // written is larger than the remaining space in the current word. Pushes
+        // all completed word into `ws`.
+        let mut write_bytes = |mut w: [u8; WORD_LENGTH], mut pos: usize, mut bytes: &[u8]| {
+            loop {
+                if bytes.len() < available::<WORD_LENGTH>(pos) {
+                    w[pos..pos + bytes.len()].copy_from_slice(bytes);
+                    return (w, pos + bytes.len());
+                } else {
+                    w[pos..].copy_from_slice(&bytes[..available::<WORD_LENGTH>(pos)]);
+                    ws.push(w);
+                    bytes = &bytes[available::<WORD_LENGTH>(pos)..];
+                    w = [0; WORD_LENGTH];
+                    pos = 0;
+                }
             }
-            let n = bytes.len() as u8;
-            let mut res = [0; WORD_LENGTH];
-            if op == Op::Insert {
-                res[0] = 1;
+        };
+
+        let mut w = [0; WORD_LENGTH];
+        let mut pos = 0;
+        for v in vs {
+            let metadata = encode_metadata(op, v.as_ref().len())?.to_be_bytes();
+            (w, pos) = write_bytes(w, pos, &metadata);
+            (w, pos) = write_bytes(w, pos, v.as_ref());
+        }
+
+        // Do not forget to push the current word if any byte were written to it.
+        if 0 != pos {
+            ws.push(w);
+        }
+
+        Ok(ws)
+    }
+
+    pub fn generic_decode<const WORD_LENGTH: usize, TryFromError: std::error::Error, Value>(
+        ws: Vec<[u8; WORD_LENGTH]>,
+    ) -> Result<HashSet<Value>, String>
+    where
+        for<'z> Value: Hash + PartialEq + Eq + TryFrom<&'z [u8], Error = TryFromError>,
+    {
+        // Attempts reading the next `n` bytes from the position `pos`.
+        let mut read_bytes = {
+            let mut ws = ws.into_iter();
+            let mut w = ws.next();
+            move |mut n: usize, pos: &mut usize| -> Option<Vec<u8>> {
+                let mut bytes = Vec::<u8>::with_capacity(n);
+                loop {
+                    if let Some(cur_w) = w {
+                        if n <= available::<WORD_LENGTH>(*pos) {
+                            cur_w[*pos..*pos + n].iter().for_each(|b| bytes.push(*b));
+                            *pos += n;
+                            return Some(bytes);
+                        } else {
+                            cur_w[*pos..].iter().for_each(|b| bytes.push(*b));
+                            n -= available::<WORD_LENGTH>(*pos);
+                            w = ws.next();
+                            *pos = 0;
+                        }
```

Comment:
why not simply ?

```suggestion
                        bytes.extend_from_slice(&cur_w[pos..pos + n]);
```

copying chunk of memory feels 'lighter' than iterating even the output is exactly same

Reply:
applied